### PR TITLE
Bug 2105967: E2E test case Telco friendly workload hints

### PR DESCRIPTION
**Pull request by mrniranjan**
Signed-off-by: Niranjan M.R <mrniranjan@redhat.com>

E2E Test Case of Telco friendly workload Hints.

### CODE REVIEW

**Comment by marioferh** [comment]
/retest-required
not sure why is failing

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 1]
```diff
@@ -566,9 +566,113 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
...
 		})
+		When("Realtime enabled", func() {
+			It("[test_id:50990][crit:high][vendor:cnf-qe@redhat.com][level:acceptance] Should updated kernel arguments and tuned", func() {
+				By("Modifying profile")
+				profile.Spec.WorkloadHints = &performancev2.WorkloadHints{
+					HighPowerConsumption: pointer.Bool(false),
+					RealTime:             pointer.Bool(true),
+				}
+				By("Updating the performance profile")
+				profiles.UpdateWithRetry(profile)
```
Comment (marioferh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:577]: I have caught up the error. The issue is because PAO doesn't detect the change in WorkloadHints, the new performanceProfile is applied but mcp does not care about it. One fast fix is also change cpus like that:
```
				By("Modifying profile")
				profile.Spec.WorkloadHints = &performancev2.WorkloadHints{
					HighPowerConsumption: pointer.Bool(false),
					RealTime:             pointer.Bool(true),
				}

				reserved := performancev2.CPUSet("0,1")
				isolated := performancev2.CPUSet("1,2")
				profile, err := profiles.GetByNodeLabels(testutils.NodeSelectorLabels)
				Expect(err).ToNot(HaveOccurred())

				profile.Spec.CPU = &performancev2.CPU{
					Reserved: &reserved,
					Isolated: &isolate
```
but also I have to investigate why mcp does not change to updating when performanceProfile changes.

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 2]
```diff
@@ -566,9 +566,112 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
...
+
+				By("Verifying that mcp is ready for update")
+				mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdated, corev1.ConditionTrue)
```
Comment (fromanirh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go]: I'm a bit concerned that hardcoding the expected sequence so stricly can be fragile. Since this will always require a reboot (will it?) perhaps better wait for `Updating`, then for `Updated` again.

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 3]
```diff
@@ -566,9 +566,112 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
...
+
+				profile.Spec.RealTimeKernel = &performancev2.RealTimeKernel{
+					Enabled: pointer.BoolPtr(false),
+				}
+
+				By("Updating the performance profile")
+				profiles.UpdateWithRetry(profile)
+
+				By("Verifying that mcp is ready for update")
+				mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdated, corev1.ConditionTrue)
```
Comment (fromanirh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:585]: same

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 4]
```diff
@@ -566,9 +566,112 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
...
```
Comment (fromanirh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:621]: same

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 5]
```diff
@@ -742,3 +839,60 @@ func getUpdatedNodes() []corev1.Node {
...
+			Expect(string(cmdline)).To(ContainSubstring(paramter))
+		}
+	}
+
+	if !rtkernel {
+		for _, node := range workerRTNodes {
+			cmd := []string{"uname", "-a"}
+			kernel, err := nodes.ExecCommandOnNode(cmd, &node)
+			Expect(err).ToNot(HaveOccurred(), "failed to execute uname")
+			Expect(kernel).To(ContainSubstring("Linux"), "Node should have Linux string")
```
Comment (fromanirh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:859]: nit: I'd write "Kernel should report itself as Linux"

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 6]
```diff
@@ -668,6 +668,7 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
 				}
 				kernelParameters := []string{rcuNocbs, "processor.max_cstate=1", "intel_idle.max_cstate=0", "intel_pstate=disable"}
 				checkWorkLoadHintTunables(workerRTNodes, stalldEnabled, sysctlMap, kernelParameters, rtKernel)
+			})
```
Comment (fromanirh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:639]: please squash with the previous commit

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 7]
```diff
@@ -566,37 +566,6 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
 			By("Saving the old performance profile")
 			initialProfile = profile.DeepCopy()
 		})
-		When("Realtime enabled", func() {
```
Comment (fromanirh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go]: so perhaps squash with the previous commit?

**Review by ffromani — COMMENTED** [review]
mostly minor comments from my side

**Review by marioferh — COMMENTED** [review]
/hold
after typo fix it seems as test are passing but it is not working right. Look at this: https://github.com/openshift/cluster-node-tuning-operator/pull/388/files
We can discuss further with @MarSik 

Also the tuned output is different with workload hint to false and no workload hints.
https://github.com/openshift/cluster-node-tuning-operator/pull/389

**Comment by mrniranjan** [comment]
@fromanirh @marioferh  Can the hold be removed ?

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 8]
```diff
@@ -566,9 +566,75 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
 			By("Saving the old performance profile")
 			initialProfile = profile.DeepCopy()
 		})
+		When("RealTime Workload with RealTime Kernel set to false", func() {
+			It("[test_id:50991][crit:high][vendor:cnf-qe@redhat.com][level:acceptance]should update kernel arguments and tuned accordingly", func() {
```
Comment (marioferh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:570]: test_id shouldn't shown, right?

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 9]
```diff
@@ -566,9 +566,75 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
...
+				By("Modifying profile")
+				profile.Spec.WorkloadHints = &performancev2.WorkloadHints{
+					HighPowerConsumption: pointer.Bool(false),
+					RealTime:             pointer.Bool(true),
+				}
+
+				profile.Spec.RealTimeKernel = &performancev2.RealTimeKernel{
```
Comment (marioferh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:577]: we are checking workloadHints, better not to add noise. If I am not missing anything its not necessary to check in this tests the value of rt kernel.

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 10]
```diff
@@ -566,9 +566,75 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
...
+				By("Updating the performance profile")
+				profiles.UpdateWithRetry(profile)
+
+				By("Applying changes in performance profile and waiting until mcp will start updating")
+				mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdating, corev1.ConditionTrue)
+
+				By("Waiting for MCP being updated")
+				mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdated, corev1.ConditionTrue)
+
+				stalldEnabled, rtKernel := true, false
```
Comment (marioferh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:590]: same here about rtkernel

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 11]
```diff
@@ -566,9 +566,75 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
...
+				noHzParam := fmt.Sprintf("nohz_full=%s", *profile.Spec.CPU.Isolated)
+				sysctlMap := map[string]string{
+					"kernel.hung_task_timeout_secs": "600",
+					"kernel.nmi_watchdog":           "0",
+					"kernel.sched_rt_runtime_us":    "-1",
```
Comment (marioferh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:595]: https://github.com/openshift/cluster-node-tuning-operator/blob/0686c0a24caffb32567dbb516d0b847cc613b933/assets/performanceprofile/tuned/openshift-node-performance#L53Missing "sched_rt_runtime_us=-1" there is kernel.sched_rt_runtime_us=-1 and

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 12]
```diff
@@ -566,9 +566,75 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
 			By("Saving the old performance profile")
 			initialProfile = profile.DeepCopy()
 		})
+		When("RealTime Workload with RealTime Kernel set to false", func() {
+			It("[test_id:50991][crit:high][vendor:cnf-qe@redhat.com][level:acceptance]should update kernel arguments and tuned accordingly", func() {
+				By("Modifying profile")
+				profile.Spec.WorkloadHints = &performancev2.WorkloadHints{
```
Comment (marioferh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:572]: HihgPowerConsumption to true and Realtime to false is the default when no workloadhints are set in performanceprofile. Bettet not to set anything here and point in the title that is the default case.

**Review by marioferh — COMMENTED** [review]
some minor comments. 
There has been tested workloadHints{false, true}{true, false}{true, true} it worth test {false, false} ?

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 13]
```diff
@@ -566,9 +566,75 @@ var _ = Describe("[rfe_id:28761][performance] Updating parameters in performance
...
+				By("Updating the performance profile")
+				profiles.UpdateWithRetry(profile)
+
+				By("Applying changes in performance profile and waiting until mcp will start updating")
+				mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdating, corev1.ConditionTrue)
+
+				By("Waiting for MCP being updated")
+				mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdated, corev1.ConditionTrue)
+
+				rcuNocbs := fmt.Sprintf("rcu_nocbs=%s", *profile.Spec.CPU.Isolated)
```
Comment (marioferh) [inline test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:623]: Why is this necessary?

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go (hunk 9 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:577]: No but in this test case we are checking workload hints with Realtime kernel disabled explicitly.

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go (hunk 10 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:590]: As mentioned above in this test case we want to check the behaviour by explicitly specifying realtime kernel to false

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go (hunk 12 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:572]: Is it not the  otherway around. "RealTime: True and Highpowerconsumption: False "

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go (hunk 8 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:570]: we add test case id to all the test cases in functional tests. As that's the only way to map the test cases in automation to test cases in our test case management system

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go (hunk 13 above)
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go]: updated in the lastest patch.

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go [hunk 14]
```diff
@@ -742,3 +802,58 @@ func getUpdatedNodes() []corev1.Node {
 	Expect(workerRTNodes).ToNot(BeEmpty(), "cannot find RT enabled worker nodes")
 	return workerRTNodes
 }
+
+//Check All tunables and kernel paramters for workloadHint
```
[the remark this answers is not in the document]
Comment (mrniranjan) [reply test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:805]: fixed in the latest commit

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go (hunk 9 above)
Comment (marioferh) [reply test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:577]: yes, but why? there is any different between workload Hints HighPower to false and realtime to true and realtime Kernel disabled explicit or even enabled?

File: test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go (hunk 12 above)
Comment (marioferh) [reply test/e2e/performanceprofile/functests/2_performance_update/updating_profile.go:572]: Yes its ok, I meant that "RealTime: True and Highpowerconsumption: False " and default, thus, with no workloadHints its the same, just to test the default behaviors. But its not necessary.

**Review by marioferh — COMMENTED** [review]
/lgtm
thanks!

**Review by ffromani — COMMENTED** [review]
/approve
